healthcheck.go

```diff
@@ -0,0 +1,117 @@
+type HealthCheckReporter struct {
+	dependencies    map[string]Dependency
+	globalHealth    bool
+	globalHealthMsg map[string]string
+	mu              sync.RWMutex
+}
+
+// HealthChecker is the interface for a Dependency that can be health-checked
+type HealthChecker interface {
+	// Check will return nil if dependency is reachable/healthy
+	// You should implement this func with a sensible timeout (< 3 or 5 sec)
+	Check() error
+}
+
+// Dependency implements the HealthChecker interfance and also contains the check config
+type Dependency struct {
+	// implements HealthChecker interface
+	HealthChecker
+	// softDep - if true it will allow errors so won't report unhealthy
+	softDep bool
+	// TODO: interval unused for now - got a global interval for all checks
+	interval time.Duration
+}
+
+// NewHealthCheckReporter returns an empty HealthCheckReporter
+func NewHealthCheckReporter() *HealthCheckReporter {
+	reporter := &HealthCheckReporter{
+		dependencies: make(map[string]Dependency),
+		globalHealth: true,
+	}
+	go func() {
+		t := time.NewTicker(5 * time.Second)
+		for {
+			<-t.C
+			reporter.RunChecks()
+		}
+	}()
+
+	return reporter
+}
+
+// RegisterDependency adds a dependency to check it periodically
+func (hr *HealthCheckReporter) RegisterDependency(name string, dep Dependency) {
+	hr.mu.Lock()
+	defer hr.mu.Unlock()
+	_, ok := hr.dependencies[name]
+	if ok {
+		panic("Dependency already exists: " + name)
+	}
+	hr.dependencies[name] = dep
+}
+
+// RunChecks performs a check in all the dependencies and update the global status
+func (hr *HealthCheckReporter) RunChecks() {
+	//create syncgroup and check all dependencies
+	var wg sync.WaitGroup
+	wg.Add(len(hr.dependencies))
+
+	depsHealth := make(map[string]string)
+	unhealthy := 0
+	for name := range hr.dependencies {
+		go func(name string) {
+			defer wg.Done()
+			//change the dependency health values
+			err := hr.dependencies[name].Check()
+			hr.mu.Lock()
+			if err != nil {
+				// increment unhealthy counter if it's a hard dependency
+				if !hr.dependencies[name].softDep {
+					unhealthy++
+				}
+				depsHealth[name] = "KO: " + err.Error()
+			} else {
+				depsHealth[name] = "OK: Healthy"
+			}
+			hr.mu.Unlock()
+		}(name)
+	}
+	// wait for all the deps to finish the checks
+	wg.Wait()
+
+	// calculate the globalHealth (if 1+ services are unhealthy, globalHealth is unhealthy too)
+	globalHealth := true
+	if unhealthy != 0 {
+		globalHealth = false
+	}
+	// refresh the calculated globalHealth
+	hr.mu.Lock()
+	hr.globalHealth = globalHealth
+	hr.globalHealthMsg = depsHealth
+	hr.mu.Unlock()
+}
+
+// Controller only has the aah HTTP Context
+type Controller struct {
+	*aah.Context
+}
+
+// HealthcheckHandler is an Aah handler to report the healthcheck
+func (c *Controller) HealthcheckHandler(hr *HealthCheckReporter) {
+	hr.mu.RLock()
+	defer hr.mu.RUnlock()
+	if hr.globalHealth == true {
+		c.Reply().Ok().JSON(hr.globalHealthMsg)
```

Comment:
even if Kubernetes only cares about the HTTP response code, I think it's nice that the endpoint always returns the JSON body with the details - it doesn't cost a lot of resources - wdyt @jeevatkm ?

Reply:
Yes, I agree let's send response always. How are you planning to handle Kubenretes liveness vs readiness routes vs non-kubernetes scenario??

Reply:
@jeevatkm 
- liveness only cares about the service being "alive" - not "healthy", so the typical scenario here is to use a primitive `/ping` endpoint which always return `200 OK - pong`. If the HTTP server is in a bad state, then this won't even work so k8s will kill the pod and start a new one.
Currently in the Aah applications that I have, I use this handler, which I'd like to add to this package, if that's OK with you too:
```
// Ping responds 200 OK pong!
func (c *Controller) Ping() {
	c.Reply().Ok().Text("%s", "pong!")
	c.Log().Trace("pong!")
}
```
- readiness should use this `healthcheck` endpoint/functionality we're implementing here, so k8s is aware that the pod is ready, and can forward requests to it.

there is a good explanation here: https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-setting-up-health-checks-with-readiness-and-liveness-probes

if dealing with a non-kubernetes scenario, I think we should keep everything as it is, and let the user or cluster-scheduler (could be Nomad, or Mesos - instead of k8s.. or could be even a VM with external HTTP healthcheck, which is very common too in GCP or AWS clouds) decide what to do with the tools we provide for this matter.

Reply:
do you agree with this @jeevatkm ?